### PR TITLE
fix: circular dependencies

### LINKED ISSUE #128: FR: Gracefully handle self-edges and circular dependencies.
Currently, we don't check for circular dependencies and self-edges. For instance `libc6` package depends on `libcrypt1` which in turn depends on `libc6` which is circular and not permitted in Bazel.

### PR DESCRIPTION
**Pull request by birunts**
Prevent adding root package as a dependency to itself.

Resolves #128 & Closes #128.

### CODE REVIEW

File: apt/private/apt_dep_resolver.bzl [hunk 1]
```diff
@@ -122,7 +122,7 @@ def _resolve_all(state, name, version, arch, include_transitive = True):
         if key in already_recursed:
             continue
 
-        if i != 0:
+        if i != 0 and key != root_package["Package"]:
```
Comment (thesayyn) [inline apt/private/apt_dep_resolver.bzl:125]: do you mind leaving a comment here that explains whats happening?

File: apt/tests/resolution_test.bzl [hunk 2]
```diff
@@ -255,10 +255,36 @@ def _resolve_aliases(ctx):
...
+resolve_circular_deps_test = unittest.make(_resolve_circular_deps_test)
+
 _TEST_SUITE_PREFIX = "package_resolution/"
 
 def resolution_tests():
     parse_depends_test(name = _TEST_SUITE_PREFIX + "parse_depends")
     resolve_optionals_test(name = _TEST_SUITE_PREFIX + "resolve_optionals")
     resolve_architecture_specific_packages_test(name = _TEST_SUITE_PREFIX + "resolve_architectures_specific")
     resolve_aliases_test(name = _TEST_SUITE_PREFIX + "resolve_aliases")
+    resolve_circular_deps_test(name = _TEST_SUITE_PREFIX + "parse_circular")
```
Comment (thesayyn) [inline apt/tests/resolution_test.bzl:290]: this technically a self-edge case, circular dependency is where a dependency somewhere endsup depending on itself.

self-edge: `libcurl -> libcurl`

circular: `libcurl -> netty -> curl -> libcurl`

Comment (birunts) [reply apt/tests/resolution_test.bzl:290]: That is what should be within the test:
`ruby -> ruby3.1 -> ruby` 

But now I've extended it to be:
`ruby -> ruby-rubygems -> ruby3.1 -> ruby`

File: apt/private/apt_dep_resolver.bzl (hunk 1 above)
Comment (birunts) [reply apt/private/apt_dep_resolver.bzl:126]: Sure :) Done.

File: apt/tests/resolution_test.bzl (hunk 2 above)
Comment (thesayyn) [reply apt/tests/resolution_test.bzl:290]: the test might pass but the generated BUILD files will still be circular afaik.

Comment (birunts) [reply apt/tests/resolution_test.bzl:327]: > but the generated BUILD files will still be circular afaik.

What exactly? Sorry I do not know the `rules_distroless` details, yet :wink: 

If it comes to generated [filegroups](https://github.com/GoogleContainerTools/rules_distroless/blob/main/apt/private/package.BUILD.tmpl#L17) (with this fix) the target won't contain itself in `srcs` attribute anymore.

Comment (birunts) [reply apt/tests/resolution_test.bzl:327]: Kindly reminder :slightly_smiling_face: @thesayyn any chance to have that reviewed/merged, please?

**Review by thesayyn — APPROVED** [review]

**Comment by thesayyn** [comment]
Turns out this did not fix the circular deps case; https://github.com/GoogleContainerTools/rules_distroless/issues/163
